src/test/librbd/test_internal.cc

```diff
@@ -610,6 +615,7 @@ TEST_F(TestInternal, SnapshotCopyup)
 
   ASSERT_EQ(256, api::Io<>::write(*ictx2, 256, bl.length(), bufferlist{bl},
                                   0));
+  ASSERT_EQ(0, flush_writeback_cache(ictx2));
```

Comment:
Why would you need to force a flush upon snapshot creation? That might break a lot of expectations if the creation of a snapshot does not flush back to the OSD.

Reply:
Withoug flushing, the data is not flushed. And making snapshots can't be done in such objects.  Currently, in the master branch, the test case passes as it doesn't enable writeback cache. I did the following test:  
Enable the writeback cache in ceph.conf  
    rbd_cache = true
    rbd_cache_writethrough_until_flush = false
    rbd_cache_policy = writeback  

And rerun the test 
RBD_FEATURES=1 ./bin/unittest_librbd --gtest_filter=TestInternal.SnapshotCopyup, it failed here: 
snap_ctx.list_snaps(ictx2->get_object_name(0), &snap_set) 

So if writeback cache is enabled, here it needs to trigger flush.

Reply:
But I guess I am still wondering why the RWL cache didn't flush prior to the creation of the snapshot? Previously it's used the `block_writes` call to ensure all IO is flushed prior to creating the snapshot. If that isn't working with RWL, it seems like it should get fixed.

Reply:
Two points here:  
1. The first point is about the test case itself.  It creates snapshots at first, block_writes() is called and ongoing IOs are flushed.  And it writes data to the original image (Line 616), and this IO doesn't been flushed if writeback cache exists. After that, it lists snapshots of the first data object. The problem is the data has been flushed to the IoCtxImpl and the first data object doesn't exist. it can't find the data object and returns an error.     
snap_ctx.list_snaps(ictx2->get_object_name(0), &snap_set)   

2. The second point is about RWL cache. The aio_flush() and flush() flush data to persistent cache devices instead of the Ceph cluster.  When creating a snapshot, it needs to close the image and create the snapshot. When closing the image, it flushes all data in the cache device to the Ceph cluster.